test/extended/openstack/loadbalancers.go

```diff
@@ -382,6 +383,152 @@ var _ = g.Describe("[sig-installer][Suite:openshift/openstack][lb][Serial] The O
 			e2e.Logf("Pod %s successfully accessed through svc loadbalancer on ip %s and port %d", podName, fip.FloatingIP, svcPort)
 		})
 	}
+
+	// Only valid with the Octavia amphora provider. The OVN-Octavia provider does not
+	// support allowed_cidrs in the LB (https://issues.redhat.com/browse/OCPBUGS-2789)
+	g.It("should limit service access on an UDP Amphora LoadBalancer when an UDP LoadBalancer svc setting the loadBalancerSourceRanges spec is created on Openshift", func() {
+
+		// Skipped for OVN-Octavia provider
+		skipIfNotLbProvider("Amphora", cloudProviderConfig)
+
+		g.By("Creating Openshift deployment")
+		depName := "udp-lb-sourceranges-dep"
+		svcName := "udp-lb-sourceranges-svc"
+		svcPort := int32(8082)
+		replicas := int32(2)
+		labels := map[string]string{"app": depName}
+		drop_sourcerange := "9.9.9.9/32"
+		testDeployment := createTestDeployment(depName, labels, replicas, v1.ProtocolUDP, 8081)
+		deployment, err := clientSet.AppsV1().Deployments(oc.Namespace()).Create(context.TODO(),
+			testDeployment, metav1.CreateOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		err = e2edeployment.WaitForDeploymentComplete(clientSet, deployment)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("Creating Openshift LoadBalancer Service with loadBalancerSourceRanges: '%s'", drop_sourcerange))
+		jig := e2eservice.NewTestJig(clientSet, oc.Namespace(), svcName)
+		jig.Labels = labels
+		svc, err := jig.CreateLoadBalancerService(loadBalancerServiceTimeout, func(svc *v1.Service) {
+			svc.Spec.Ports = []v1.ServicePort{{Protocol: v1.ProtocolUDP, Port: svcPort, TargetPort: intstr.FromInt(8081)}}
+			svc.Spec.Selector = labels
+			svc.Spec.LoadBalancerSourceRanges = []string{drop_sourcerange}
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("Checks from openshift perspective")
+		loadBalancerId := svc.GetAnnotations()["loadbalancer.openstack.org/load-balancer-id"]
+		o.Expect(loadBalancerId).ShouldNot(o.BeEmpty(), "load-balancer-id annotation missing")
+		e2e.Logf("LB ID: %s", loadBalancerId)
+		o.Expect(svc.Spec.LoadBalancerSourceRanges).Should(o.Equal([]string{drop_sourcerange}))
+		svcIp := svc.Status.LoadBalancer.Ingress[0].IP
+		o.Expect(svcIp).ShouldNot(o.BeEmpty(), "FIP missing on svc Status")
+		e2e.Logf("Service IP: %s", svcIp)
+
+		g.By("Checks from openstack perspective")
+		allPages, err := octavialisteners.List(loadBalancerClient, octavialisteners.ListOpts{LoadbalancerID: loadBalancerId}).AllPages()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		listeners, err := octavialisteners.ExtractListeners(allPages)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// Check there is only one listener in the LB
+		o.Expect(listeners).Should(o.HaveLen(1), "More than one listener found for LB '%s': %v", loadBalancerId, listeners)
```

Comment:
What if it's 0? We should change the log here.

Reply:
Right, I've changed it to:
```
o.Expect(listeners).Should(o.HaveLen(1), "Unexpected number of listeners for LB '%s'", loadBalancerId)
```
The error log will show the listeners object and the expected length:
i.e (ignore the values as I changed the expected number of listeners in order to trigger the error):

```
fail [/home/stack/openstack-test/test/extended/openstack/loadbalancers.go:452]: Unexpected number of listeners for LB '36305d21-e3dc-4870-84f3-f4cdcc872b13'
Expected
    <[]listeners.Listener | len:1, cap:4>: [
        {
            ID: "a027f57d-4c12-496e-80e7-5d156105d3ab",
            ProjectID: "30d0f65255d94819b724e76a8f81e0ea",
            Name: "listener_0_kube_service_kubernetes_e2e-test-openstack-qczsq_udp-lb-sourceranges-svc",
            Description: "",
            Protocol: "UDP",
            ProtocolPort: 8082,
            DefaultPoolID: "ffe26c35-d849-4238-be41-325045848cf8",
            Loadbalancers: [
                {
                    ID: "36305d21-e3dc-4870-84f3-f4cdcc872b13",
                },
            ],
            ConnLimit: -1,
            SniContainerRefs: [],
            DefaultTlsContainerRef: "",
            AdminStateUp: true,
            Pools: nil,
            L7Policies: [],
            ProvisioningStatus: "ACTIVE",
            TimeoutClientData: 50000,
            TimeoutMemberData: 50000,
            TimeoutMemberConnect: 5000,
            TimeoutTCPInspect: 0,
            InsertHeaders: {},
            AllowedCIDRs: ["9.9.9.9/32"],
        },
    ]
to have length 2
```